### PR TITLE
Add incarnation ID to taskQueueManager fatal signal

### PR DESCRIPTION
**Pull request by mmcshane**
**What changed?**
A taskQueueManager shutting down can result in multiple upcalls to
unload the queue on different goroutines. We want only the first of
those calls to succeed and for later calls to have no effect. If the
later calls are allowed to run, they will unload a successor to the
taskQueueManager that emitted the signal in the first place.

Also uses github.com/google/uuid for UUIDs as this library is
maintained and more featureful than the pborman one. In fact, it's
maintained by pborman.

**Why?**
With enough activity and failure, the fatal signals can become self-sustaining

**How did you test it?**
New unit tests

**Potential risks**
Minimal. The unload call now performs an unload in strictly _fewer_ scenarios.

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/matching/taskQueueManager.go [hunk 1]
```diff
@@ -107,6 +98,10 @@ type (
 		GetAllPollerInfo() []*taskqueuepb.PollerInfo
 		// DescribeTaskQueue returns information about the target task queue
 		DescribeTaskQueue(includeTaskQueueStatus bool) *matchingservice.DescribeTaskQueueResponse
+		// Observe this object's incarnation identity. Many taskQueueManager
+		// objects can serve a given task queue over time; each is uniquely
+		// identified.
+		IncarnationID() uuid.UUID
```
Comment (mmcshane) [inline service/matching/taskQueueManager.go:104]: I would really prefer for the taskQueueManager to be unaware of its own incarnation information. IMO that information should be tracked by the entity that is in charge of the taskQueueManager's lifecycle, viz., the matchingEngine. However the several of the matchingEngine tests reach into the implementation `map[taskQueueID]taskQueueManager` and swap out the `taskQueueManager` instance with a new one. Sometimes this new instance has been started, sometimes not. 

tldr: this is in the wrong place because we have some bad practices in the unit tests and we should stop mucking about with private internals. Using `thing_test` packages would be a good start.

Comment (yiminc) [reply service/matching/taskQueueManager.go:104]: I thought we always re-create the taskQueueManager when it reload? If that is true, why we cannot compare if the instance request unload is the same instance in the map hold by matching engine? We don't need this new UUID, but just need to check if the current instance from matching engine's map is the same instance requesting to unload. Would that work?

Comment (mmcshane) [reply service/matching/taskQueueManager.go:104]: That's basically what the incarnation ID does. I wasn't confident that the value of the pointer would definitely change as Go doesn't really make promises about that. Or were you thinking of some other identifier?

Comment (yiminc) [reply service/matching/taskQueueManager.go:104]: We could just use the interface.
signalFatalProblem func(taskQueueManager)
Then you pass in pointer of the current taskQueueMangerImpl, and in matchingEngine, just check if it is still the same instance in the map.

**Comment by mmcshane** [comment]
#1917 instead
